### ISSUE TITLE
Post-841: Split Stream

### ISSUE THREAD

**Issue by phadej** [haskell-servant/servant#995]
Currently we have `Verb` and `Stream`. After #841 refactor we'll have `Result` and `Stream`. So far so good.

Yet when considering the implementation of #841, and also serfvant-docs problem in #989 I think we need to do following:

*Split `Stream` into `FramedResults` and `StreamedResult` (aka `NoFraming`).*

Note the singular vs. plural. `StreamedResult` is still single result, the implementation of server and clients is streaming because the thing might be huge. `FramedResults` is plural: many (probably small) results in streaming fashion.

For `StreamedResult` we'll need a new pair of type-classes: `MimeStream` / `MimeUnstream` (compare to `MimeRender` / `MimeUnrender` and `ToStreamGenerator` / `FromResultStream`).

ping @gbaz, @jvanbruegge, @alpmestan, @jkarni

**Comment by jvanbruegge** [comment]
Could you explain *why* you think that `StreamedResults` would have to be split? And second: How would having a single `StreamedResult` look like?

**Comment by phadej** [comment]
```haskell
type API = "verybigfile" :> StreamedResult JSON Document -- single big thing
    :<|> "alotoflines" :> FramedResult NewlineFraming 'PlainText [Text]
```

Note that with previous type:

```haskell
toStreamGenerator :: f a -> StreamGenerator a
```

the problem was that `Conduit i o m r` has arguments in "wrong" order. (`a = o`).

It's visibile from `FramingUnrender NoFraming` instance:

https://github.com/haskell-servant/servant/blob/77ea599c63efbd9110b693e1f4e40aa4b98f8a10/servant/src/Servant/API/Stream.hs#L95-L97

There are no chunking into frames at all. `servant-client` using `Stream NoFraming` will behave *very bad*. Currently, for example `StreamGet NoFraming [x]` makes only sense to the types `x` which `MimeUnrender` never fails (essentially `ByteString`).

---

There is e.g. http://ndjson.org/ which can be produced with `StreamGet NewlineFraming JSON [a]`. But then the result is not `application/json`! Here servant is clearly broken. For framed stuff, we need to mangle MIME types!

---

All above a small smells here and there. We need to separate
- streaming single big thing
- returning multiple things framed (*note*: that doesn't even need to be streamed, it makes sense though)

**Comment by alpmestan** [comment]
Right, I think this was the outcome of our IRC discussions a while ago.

> we'll need a new pair of type-classes: `MimeStream` / `MimeUnstream`

I'm a little less enthusiastic about this. But the alternative is probably to change `MimeRender`/`MimeUnrender` to be able to accomodate both bytestring responses and streams? Doesn't sound great either for now, as that'll break an _insane_ amount of code.

**Comment by gbaz** [comment]
I don't have the context in my head for those discussions anymore. For the docs stuff (which is the same as just getting the content-type correct) we need a new typeclass that derives the content type out of the _combination_ of the framing strategy and carried data. (There's not just "one way" to frame any given underlying type of data -- ndjson and netstring are both valid ways to transmit a sequence of json, for example). That said, I don't think that the introduction of this typeclass involves splitting streaming types or not either way.

So I think that leaves this comment: "servant-client using Stream NoFraming will behave very bad. Currently, for example StreamGet NoFraming [x] makes only sense to the types x which MimeUnrender never fails (essentially ByteString)," which I'm afraid I still don't follow.

`StreamGet NoFraming [x]` should be no different than `Get [x]` imho? Without framing for boundary recovery, there's no reason to expect that it should be any different than getting any other thing where if it cuts off partway everything goes *splunk*, unless everything goes *splunk* in a much worse way here? And if it does, can we just fix that in the code?

**Comment by phadej** [comment]
`NoFraming` is not wrong thing to have. It's not related to returning multiple results

Look at the motivation of @jvanbruegge https://github.com/haskell-servant/servant/pull/959#issuecomment-391156243 example

```haskell
type FileStream = ConduitT () ByteString (ResourceT IO) ()

serveFile :: Id -> Server (StreamGet NoFraming OggVideo FileStream)
serveFile id = except404 (getFilePath id) >>= pure . Conduit.sourceF
```

It serves single *file*. You can think about it as a collection of chunks, but it's not. I argue that `servant` way is to have:

```haskell
data VideoFile = VideoFile { vfMetadata :: MetaData, vfData :: ... }

type API = StreamedResult '[OggVideo] VideoFile

instance MimeStreamRender OggVideo VideoFile where ...
```

---

Another way to think is to think it from client POV for `StreamGet NoFraming [x]` how conversion goes:

```haskell
InputData                -- Stream instance
PseudoStream ByteString  -- using FramingUnrender (PseudoStream is non-existing, implicit in how Stream instance is written)
ResultStream x           -- BuildFromStream, using `MimeUnrender x` instance
                         -- note: not MimeUnrender x!
[x]
```

We use `MimeUnrender x`, not `[x]` so we'll get completely wrong result.
In theory `NoFraming JSON Value` input such that:

```
[][1][2][3][4]
```

could be splittable in frames, but then we'll need to have an incremental parser
in `MimeUnrender` (and that's complication I'm not happy to tackle).

---

It's easy to write (and I will) `MimeRender` / `MimeUnrender` instances such that:

```haskell
type API = Get '[FramedContent Netstring JSON] [Int]
```

Note that

```haskell
Get '[FramedContent NoFraming JSON] [ByteString]
```

is weird thing to have. You could just have `Lazy.ByteString`.


---

TL;DR Framing and Streaming are orhogonal concepts, and they should be so 
in servant.

**Comment by gbaz** [comment]
Ok. You suggest `StreamedResult '[OggVideo] VideoFile` instead of `StreamGet NoFraming OggVideo FileStream`, but modulo having a special type for `VideoFile` instead of `FileStream` isn't this basically equivalent to `type StreamedResult  = StreamGet NoFraming`? We can currently implement that special type not through any need for an explicit `MimeStreamRender` but just by giving a `VideoFile` a `ToStreamGenerator` instance directly, no?

The reason framing and streaming are not orthogonal is that the streaming is performed in chunks, and the chunks interact with streaming, because we stream one framed chunk at a time. There is no reason to frame data rather than just using a different mime type unless we take advantage of the chunkwise character of streaming, and you certainly can have streaming without framing, but it is just the straightforward case.

That said, I do now see your point that `NoFraming` as a framing strategy now currently doesn't make sense with certain parsers, and in particular that `StreamGet NoFraming 'JSON [Int]` for example is weird. But if we have a `StreamMimeType` class or whatever, then that helps resolve that, since there could just be no instance given (or an error instance) for things like `NoFraming 'JSON` that don't make sense, and there could be an `NDJSON` instance given for `NewlineFraming 'JSON`, etc.

**Comment by jvanbruegge** [comment]
I would also agree with @gbaz here. Framing is an extension of streaming, not orthogonal. Also some framing strategies dont make sense for binary data, while others do (something like prepending the size of the chunk before each chunk, had to build a protocol like this once at work a while ago). And as @phadej said, framing may change the content type of your response. So I see the case for the two parameter type class here.
But I disagree that streaming binary data (aka a file) is the same as a lazy bytestring (if this is what you meant). Streaming gives you a lot more control over the resource handling and memory usage.

**Comment by jvanbruegge** [comment]
@phadej isnt the definition of "framing" literally to enclose a chunk with extra data (ie *frame* it)? That's what I was assuming, so in your example, only the second one would be valid, because it is a set of chunks that is framed.

**Comment by gbaz** [comment]
@phadej yeah I tend to agree with @jvanbruegge here. I think we have a significant disconnect in how things are currently setup and the purpose. Framing is a strategy to deliniate chunks. `NewlineFraming` is a bad example here because it has an invariant that your examples with `ByteString` all violate. From the haddocks: "This assumes that it is used with a Content-Type that encodes without newlines (e.g. JSON)." In any case it _creates_ a frame for each item in the list. There is no distinction between chunks and frames -- each chunk (i.e. item in the list or other enumeration) is converted into a frame.

You are correct that `instance ToStreamGenerator VideoFile ByteString` ignores the mime type. But that's what I'm proposing we change -- we derive the mimetype from the _combination_ of the data carried and the streamtype. So we'd have `instance StreamMimeType NoFraming VideoFile` to yield the correct mime type.

I am very much trying to bear client needs in mind here as well. When I wrote the initial streaming stuff I made sure to have it work also with the client, and wrote all the client-side code to make it go. At the time, I realized the whole mimetype stuff was an infelicity we'd have to fix later. So what I'm proposing is an alternate approach to fixing it that I think more accurately captures what's going on.

Note that in the first example, by _not having _ an `instance StreamMimeType NewlineFraming ByteString` we could prevent the bad code, or we could even add an error instance in such a case to definitively steer people away from it.
